qemu/tests/ansible_test.py

```diff
@@ -0,0 +1,83 @@
+    if not (sm.check_installed("ansible") or sm.install("ansible")):
+        test.cancel("ansible package install failed")
+
+    get_ip_timeout = params.get_numeric("get_ip_timeout")
+    guest_user = params["username"]
+    guest_passwd = params["password"]
+    ansible_callback_plugin = params["ansible_callback_plugin"]
+    ansible_addl_opts = params.get("ansible_addl_opts", "")
+    ansible_ssh_extra_args = params["ansible_ssh_extra_args"]
+    ansible_extra_vars = params.get("ansible_extra_vars", "{}")
+    playbook_repo = params["playbook_repo"]
+    playbook_timeout = params.get_numeric("playbook_timeout")
+    playbook_dir = params.get("playbook_dir",
+                              os.path.join(test.logdir, "ansible_playbook"))
+    tl_playbook = os.path.join(playbook_dir, params["tl_playbook"])
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    guest_ip = vm.wait_for_get_address(0, get_ip_timeout)
+
+    try:
+        logging.info("Clone %s", playbook_repo)
+        process.run("git clone {src} {dst}".format(src=playbook_repo,
+                                                   dst=playbook_dir),
+                    verbose=False)
+
+        error_context.base_context("Generate playbook related options",
+                                   logging.info)
+        extra_vars = {"ansible_ssh_extra_args": ansible_ssh_extra_args,
+                      "ansible_ssh_pass": guest_passwd}
+        extra_vars.update(json.loads(ansible_extra_vars))
+
+        ansible_cmd_options = ["ansible-playbook",
+                               "-u {}".format(guest_user),
+                               "-i {},".format(guest_ip),
+                               "-e '{}'".format(json.dumps(extra_vars)),
+                               ansible_addl_opts,
+                               tl_playbook]
+        ansible_cmd = r" ".join(ansible_cmd_options)
+
+        error_context.context("Execute ansible playbook", logging.info)
+        env_vars = {"ANSIBLE_STDOUT_CALLBACK": ansible_callback_plugin}
+        play_s, play_o = process.getstatusoutput(ansible_cmd,
+                                                 timeout=playbook_timeout,
+                                                 ignore_status=False,
+                                                 shell=False, env=env_vars)
+        logging.debug("Full ansible playbook output for debugging:\n%s",
+                      play_o)
+        if play_s != 0:
+            test.fail("Ansible playbook execution failed, please check the "
+                      "log for details.")
+        logging.info("Ansible playbook execution passed.")
+    except process.CmdError as err:
+        logging.error("Test terminated unexpectedly!\nStderr:%s\nStdout:%s",
+                      err.result.stderr_text, err.result.stdout_text)
+        test.error("Unexpected error during executing ansible playbook test.")
+    finally:
+        shutil.rmtree(playbook_dir, ignore_errors=True)
```

Comment:
Can replace the `test.logdir` with `test.workdir`([doc](https://avocado-framework.readthedocs.io/en/79.0/api/test/avocado.html?highlight=workdir#avocado.Test.workdir)) for `playbook_dir`, then seems not need to remove the temp dir here.

Reply:
I set it to `test.logdir` is in order to copy the log file from the remote host, the `test.logdir` can be the default if I missed setting the host_log_dir in the YAML file and the `test.workdir` will be cleaned up once the test finishes.
```yaml
host_log_dir: '{{ playbook_dir|dirname }}'
```
@luckyh Do you have a better idea?

Reply:
Better to distinguish these two directories, we should make the log dir as an interface to all playbooks, that means if playbooks want to copy log files back to host they can only take it as the target (just likes an ENV VAR). And we can make playbooks always be cloned into `test.workdir`, and can introduce a new param (e.g. `workspace_dir`) if users do need to specify the clone target.